app/utilities/repository_import_parser/importer.rb

```diff
@@ -0,0 +1,130 @@
+      @header_skipped = false
+      @repository = repository
+      @sheet = sheet
+      @rows = SpreadsheetParser.spreadsheet_enumerator(@sheet)
+      @mappings = mappings
+      @user = user
+      @repository_columns = @repository.repository_columns
+    end
+
+    def run
+      get_columns
+      return check_for_duplicate_columns if check_for_duplicate_columns
+      import_rows!
+    end
+
+    private
+
+    def get_columns
+      @mappings.each.with_index do |(_, value), index|
+        if value == '-1'
+          # Fill blank space, so our indices stay the same
+          @columns << nil
+          @name_index = index
+        else
+          @columns << @repository_columns.find_by_id(value)
+        end
+      end
+    end
+
+    def check_for_duplicate_columns
+      col_compact = @columns.compact
+      if col_compact.map(&:id).uniq.length != col_compact.length
+        return { status: :error,
+                 nr_of_added: @new_rows_added,
+                 total_nr: @total_new_rows }
+      end
+    end
+
+    def import_rows!
+      errors = false
+      @rows.each do |row|
+        # Skip empty rows
+        next if row.empty?
+        unless @header_skipped
+          @header_skipped = true
+          next
+        end
+        @total_new_rows += 1
+
+        row = SpreadsheetParser.parse_row(row, @sheet)
+        record_row = new_repository_row(row)
+        record_row.transaction do
+          unless record_row.save
+            errors = true
+            raise ActiveRecord::Rollback
+          end
+
+          row_cell_values = []
+
+          row.each.with_index do |value, index|
+            column = @columns[index]
+            if column && value.present?
+              # uses RepositoryCellValueResolver to retrieve the correct value
+              cell_value_resolver =
+                RepositoryImportParser::RepositoryCellValueResolver.new(
+                  column, @user, @repository
+                )
+              cell_value = cell_value_resolver.get_value(value, record_row)
+              unless cell_value.valid?
+                errors = true
+                raise ActiveRecord::Rollback
+              end
+              row_cell_values << cell_value
+            end
+          end
+
+          unless import_to_database(row_cell_values)
+            errors = true
+            raise ActiveRecord::Rollback
+          end
+          @new_rows_added += 1
+        end
+      end
+
+      if errors
+        return { status: :error,
+                 nr_of_added: @new_rows_added,
+                 total_nr: @total_new_rows }
+      end
+      { status: :ok, nr_of_added: @new_rows_added, total_nr: @total_new_rows }
+    end
+
+    def new_repository_row(row)
+      RepositoryRow.new(name: row[@name_index],
+                        repository: @repository,
+                        created_by: @user,
+                        last_modified_by: @user)
+    end
+
+    def import_to_database(row_cell_values)
```

Comment:
How can we extend this later?

Reply:
you make a switch statement or I don't know do we need this to be extendable from addons?

Reply:
Yes, to add a data type from addon